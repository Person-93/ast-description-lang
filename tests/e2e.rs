use anyhow::{ensure, Context, Result};
use escargot::CargoBuild;
use insta::{assert_snapshot, with_settings};
use std::{env, fs, process::Stdio};

fn main() -> Result<()> {
  for entry in fs::read_dir("./examples").context("failed to read examples directory")? {
    let entry = entry.context("failed to read directory entry")?;
    if !entry.metadata()?.is_dir() {
      continue;
    }
    let path = entry.path();
    println!("running example at {}", path.display());
    let output = CargoBuild::new()
      .manifest_path(path.join("Cargo.toml"))
      .run()?
      .command()
      .stdout(Stdio::piped())
      .spawn()
      .context("failed to spawn process for example")?
      .wait_with_output()
      .context("failed waiting for output from example")?;
    ensure!(
      output.status.success(),
      "example-{} exited non zero",
      path.file_name().unwrap().to_str().unwrap()
    );
    with_settings!({input_file => Some(path)}, {
      assert_snapshot!(std::str::from_utf8(&output.stdout).unwrap())
    });
  }

  Ok(())
}
