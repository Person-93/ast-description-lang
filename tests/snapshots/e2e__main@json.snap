---
source: tests/e2e.rs
assertion_line: 67
expression: "std::str::from_utf8(&output.stdout).unwrap()"
input_file: examples/json

---
Object(
    [
        Member {
            str_lit: StrLit(
                "\"foo\"",
            ),
            json: StrLit(
                StrLit(
                    "\"bar\"",
                ),
            ),
        },
        Member {
            str_lit: StrLit(
                "\"obj\"",
            ),
            json: Object(
                [
                    Member {
                        str_lit: StrLit(
                            "\"answer\"",
                        ),
                        json: NumLit(
                            NumLit(
                                42,
                                None,
                            ),
                        ),
                    },
                ],
            ),
        },
        Member {
            str_lit: StrLit(
                "\"arr\"",
            ),
            json: Array(
                [
                    NumLit(
                        NumLit(
                            2,
                            None,
                        ),
                    ),
                    NumLit(
                        NumLit(
                            3,
                            None,
                        ),
                    ),
                    NumLit(
                        NumLit(
                            4,
                            None,
                        ),
                    ),
                    StrLit(
                        StrLit(
                            "\"hi\"",
                        ),
                    ),
                ],
            ),
        },
    ],
)
